1-Draft/RFCNNNN-Improve-Background-Operator.md

```diff
@@ -0,0 +1,216 @@
+
+```powershell
+# This command:
+Get-Process &!
+# Would return the same output as this command:
+Start-Job {Get-Process} -ShowOutput
+```
+
+```output
+Id     Name            PSJobTypeName   State         HasMoreData     Location             Command
+--     ----            -------------   -----         -----------     --------             -------
+1      Job1            BackgroundJob   Running       True            localhost            Microsoft.PowerShell.Man…
+```
+
+In addition to the output above, which is the job object created by the command
+that was invoked, PowerShell would show the output from the job (and any other
+job) launched this way in an overlay much like the output from Write-Progress,
+with the job output from any stream appearing as it is output in the overlay,
+prefixed with "[JobName]:" (e.g. [Job1]: ). This overlay would be slightly
+larger than Write-Progress so that users could see more output from the job in
+the overlay, and when all jobs sharing output complete the overlay would
+disappear after a delay (5 seconds). Progress messages from the jobs would
+appear above the output.
+
+### Launching a background job quietly and referencing the job with the `$!` variable
+
+```powershell
+# These commands:
+$null = Get-Process &
+$!
+# Would return the same output as these commands:
+$null = Start-Job {Get-Process}
+$!
+```
+
+```output
+Id     Name            PSJobTypeName   State         HasMoreData     Location             Command
+--     ----            -------------   -----         -----------     --------             -------
+1      Job1            BackgroundJob   Running       True            localhost            Microsoft.PowerShell.Man…
+```
+
+The `&` operator permits launching a job using an operator without any of the
+messages from the various streams appearing by default in the console (this is
+how this operator works today).
+
+The `$!` variable in the subsequent command returns the job that was created
+with the command before it.
+
+### Fanning out to multiple background jobs and referencing the jobs with the `$!` variable
+
+```powershell
+$null = cmd1 & cmd2 & cmd3 &
+$!
+```
+
+```output
+Id     Name            PSJobTypeName   State         HasMoreData     Location             Command
+--     ----            -------------   -----         -----------     --------             -------
+1      Job1            BackgroundJob   Running       True            localhost            Microsoft.PowerShell.Man...
+2      Job1            BackgroundJob   Running       True            localhost            Microsoft.PowerShell.Man...
+3      Job1            BackgroundJob   Running       True            localhost            Microsoft.PowerShell.Man...
+```
+
+The `$!` variable in those two commands returns the jobs that were created with
+the command before it.
+
+### Launching a legacy command with the stop-parsing sigil and either the `&` or `&!` background operator
```

Comment:
I just spent a while trying to figure out if this could break something. On Windows, I *think* we're okay given that `&` is a

But I've also been learning a little bit more about our parser and lookahead lately, and I'm curious if there's anything from a parsing perspective that would make this tricky. (I'm assuming that the `&` or `&!` would have to be the last character on the line for this to work properly). It also kind of breaks the contract with what `--%` is supposed to do, but I get why this might make sense as an exception (assuming the implementation isn't really hard). 

I think it's also worth calling out here that there are existing workarounds for this (neither of which I think are particularly hairy, especially if we documented them):

```powershell
# These are equivalent

$a = foo.exe --% /arg1 val1 /arg2 val2
& $a &

& {
    foo.exe --% /arg1 val1 /arg2 val2
}
```

On Linux, we don't actually support `--%` so it's not relevant there (per PowerShell/PowerShell#3733)

Reply:
I don't think your example commands are written quite the way you intended, but I get the point. That's fair. I thought it might be useful if `&` (and `&!` if it gets implemented) terminate the stop-parsing mode since they are statement terminators, but it was only added for completeness. Since the workaround is easy, since it's not a must have, and since it's Windows-only, that can be removed.

Reply:
Hahaha yeah, they were both wrong, edited them in a discussion yesterday, apologies. 

Sounds good.

Reply:
Removed the suggestions around the stop-parsing sigil.